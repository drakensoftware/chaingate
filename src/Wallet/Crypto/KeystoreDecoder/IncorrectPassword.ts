export default class IncorrectPassword extends Error {
    constructor() {
        super('Password provided is not correct')
        if (Error.captureStackTrace) Error.captureStackTrace(this, IncorrectPassword)
        this.name = this.constructor.name
    }
}
