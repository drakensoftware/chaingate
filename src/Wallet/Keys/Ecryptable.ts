import {Utils} from '../../Utils'
import {decryptAES, encryptAES, EncryptedData} from '../Crypto/AES'
import {AsyncMutex} from '@esfx/async-mutex'

class EncryptionError extends Error {
    constructor(message: string) {
        super(message)
        if (Error.captureStackTrace) Error.captureStackTrace(this, EncryptionError)
        this.name = this.constructor.name
    }
}

export class WalletIsEncrypted extends Error {
    constructor() {
        super('Wallet is currently encrypted. ' +
            'You should run functions that require the usage of private key with runUnencrypted(password, ...)')
        if (Error.captureStackTrace) Error.captureStackTrace(this, WalletIsEncrypted)
        this.name = this.constructor.name
    }
}

export abstract class Encryptable {
    private data: Uint8Array | EncryptedData
    private runDecryptedMuxed = new AsyncMutex()
    public enableRunningUnencryptedMessage = false

    get isEncrypted(): boolean {
        return this.data instanceof EncryptedData
    }

    get raw(): Uint8Array{
        if(this.data instanceof EncryptedData) throw new WalletIsEncrypted()
        if(this.enableRunningUnencryptedMessage && !this.runDecryptedMuxed.isLocked && !process.env.I_AM_SURE_I_AM_NOT_IN_PRODUCTION)
            console.warn('You are running with a wallet unencrypted. This should only be done for development purposes.\n' +
                'If you intend to use it in production, call encrypt(password) after it is created,\n' +
                'and run functions that require the usage of private key with runUnencrypted(password, ...)')
        return this.data
    }

    get hexa(){
        return Utils.bytesToHex(this.raw, false)
    }

    constructor(rawData: Uint8Array){
        this.data = rawData
    }

    async encrypt(password: string){
        if(this.data instanceof EncryptedData) throw new EncryptionError('Wallet is already encrypted')
        this.data = await encryptAES(this.data, password)
    }

    private async decrypt(password: string){
        if(!(this.data instanceof EncryptedData)) throw new EncryptionError('Wallet is not encrypted')
        this.data = await decryptAES(this.data, password)
    }

    async runUnencrypted<T>(password: string, fn: () => Promise<T>): Promise<T> {
        if(!this.isEncrypted) throw new Error('Wallet is not encrypted')
        try {
            await this.runDecryptedMuxed.lock()
            await this.decrypt(password)
            const result = await fn() // Capture the result of fn to return later
            await this.encrypt(password)
            return result // Return the result after locking
        } finally {
            this.runDecryptedMuxed.unlock() // Ensure mutex is always unlocked
        }
    }
}
