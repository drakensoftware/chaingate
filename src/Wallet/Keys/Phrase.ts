import {Encryptable, Seed} from './index'
import {wordLists} from '../../Utils/PhraseGenerator'
import * as bip39 from '@scure/bip39'
import {HDKey} from '@scure/bip32'
import {mnemonicToSeed} from '@scure/bip39'

export class Phrase extends Encryptable{
    constructor(phrase: string | Uint8Array){
        if(phrase instanceof Uint8Array) super(phrase)
        else super(new TextEncoder().encode(phrase))
    }

    async derive(derivationPath: string){
        return (await this.getSeed()).derive(derivationPath)
    }

    async getSeed(){
        return new Seed(await mnemonicToSeed(this.phrase))
    }

    get phrase(){
        return new TextDecoder().decode(this.raw)
    }

    static isValidPhrase(phrase: string): boolean{
        for(const wordlist of Array.from(wordLists.values()))
            if(bip39.validateMnemonic(phrase, wordlist)) return true
        return false
    }
}
