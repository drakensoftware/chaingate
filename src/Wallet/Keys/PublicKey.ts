export class PublicKey {
    readonly compressed: Uint8Array
    readonly uncompressed: Uint8Array


    constructor(compressed: Uint8Array, uncompressed: Uint8Array){
        this.compressed = compressed
        this.uncompressed = uncompressed
    }

}
