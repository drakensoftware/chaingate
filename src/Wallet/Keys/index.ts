export {Encryptable} from './Ecryptable'
export {PublicKey} from './PublicKey'
export {PrivateKey} from './PrivateKey'
export {Seed} from './Seed'
export {Phrase} from './Phrase'
export {PrivateKeySource} from './PrivateKeySource'
