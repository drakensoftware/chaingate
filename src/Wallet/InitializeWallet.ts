import {PhaseNumOfWords, PhraseGenerator, PhraseLanguage} from '../Utils/PhraseGenerator'
import {PhraseWallet} from './PhraseWallet'
import {ChainGateClient} from 'chaingate-client'
import {Phrase} from './Keys'
import {decryptKeystoreV1, isKeystoreV1} from './Crypto/KeystoreDecoder/KeystoreV1'
import {decryptKeystore} from './Crypto/KeystoreDecoder/OtherKeystores'
import {ImportedPrivateKey} from './ImportedPrivateKey'
import {PrivateKey} from './Keys'
import IncorrectPassword from './Crypto/KeystoreDecoder/IncorrectPassword'
import {Utils} from '../Utils'
import {HDWallet} from './HDWallet'
import {Seed} from './Keys'
import wif from 'wif'

export class EncodingError extends Error {
    constructor(message: string) {
        super(message)
        if (Error.captureStackTrace) Error.captureStackTrace(this, EncodingError)
        this.name = this.constructor.name
    }
}

export async function create(apiKey: string,
    phraseLanguage: PhraseLanguage = 'english', phraseNumOfWords: PhaseNumOfWords = 12) {
    const phrase = PhraseGenerator.generateNewPhrase(phraseLanguage, phraseNumOfWords)
    return await PhraseWallet.new(new ChainGateClient(apiKey), new Phrase(phrase))
}

export async function fromKeystore(apiKey: string, keystore: string, password: string) {
    //Some keystores store the phrase and others the private key
    try {
        const decodedKeystore = isKeystoreV1(keystore)
            ? await decryptKeystoreV1(keystore, password)
            : await decryptKeystore(keystore, password)

        let phrase
        try {
            phrase = new TextDecoder().decode(decodedKeystore)
        } catch { /* empty */ }

        if (phrase && Phrase.isValidPhrase(phrase)) {
            return PhraseWallet.new(new ChainGateClient(apiKey), new Phrase(phrase))
        } else {
            return ImportedPrivateKey.new(new ChainGateClient(apiKey), new PrivateKey(decodedKeystore))
        }
    } catch (ex) {
        if (ex instanceof IncorrectPassword) throw ex
        throw new EncodingError('Invalid keystore')
    }
}

export async function fromPhrase(apiKey: string, phrase: string) {
    if(!Phrase.isValidPhrase(phrase)) throw new EncodingError('Invalid phrase')
    return PhraseWallet.new(new ChainGateClient(apiKey), new Phrase(phrase))
}

export async function fromSeed(apiKey: string, seed: string | Uint8Array, ){
    if(typeof seed === 'string' && Utils.isHex(seed)) seed = Utils.hexToBytes(seed)
    else if(typeof seed === 'string') throw new EncodingError('Invalid seed')

    return HDWallet.new(new ChainGateClient(apiKey), new Seed(seed))
}

export async function  fromPrivateKey(apiKey: string, privateKey: string | Uint8Array){
    if(typeof privateKey === 'string' && Utils.isHex(privateKey)) privateKey = Utils.hexToBytes(privateKey)
    else if(typeof privateKey === 'string' && Utils.isBase58(privateKey)){
        try{
            privateKey = new Uint8Array(wif.decode(privateKey).privateKey)
        }catch (_ex){
            throw new EncodingError('The string supplied in Wallet Import Format (WIF) is deemed to be invalid')
        }
    }else if(typeof privateKey === 'string') throw new EncodingError('Private key is invalid or the format is unrecognized')

    return ImportedPrivateKey.new(new ChainGateClient(apiKey), new PrivateKey(privateKey))
}
