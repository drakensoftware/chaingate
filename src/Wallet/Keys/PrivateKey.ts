import wif from 'wif'
import {secp256k1} from '@noble/curves/secp256k1'
import {PublicKey} from './PublicKey'
import {Encryptable} from './Ecryptable'

export class PrivateKey extends Encryptable {
    get wif(){
        return wif.encode(128, Buffer.from(this.raw), true) //128 is bitcoin mainnet
    }

    async getPublicKey(): Promise<PublicKey>{
        return new PublicKey(
            secp256k1.getPublicKey(this.raw, true),
            secp256k1.getPublicKey(this.raw, false)
        )
    }
}
