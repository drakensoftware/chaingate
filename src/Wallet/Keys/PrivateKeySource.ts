import {PrivateKey} from './PrivateKey'
import {Phrase} from './Phrase'
import {Seed} from './Seed'

export type PrivateKeySource = Phrase | Seed | PrivateKey
