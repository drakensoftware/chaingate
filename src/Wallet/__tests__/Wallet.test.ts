import IncorrectPassword from '../Crypto/KeystoreDecoder/IncorrectPassword'
import {EncodingError} from '../InitializeWallet'
import * as InitializeWallet from '../InitializeWallet'
import {WalletIsEncrypted} from '../Keys/Ecryptable'

beforeEach(() => {
    jest.spyOn(global.crypto, 'getRandomValues').mockImplementation((buffer) => {
        const n = buffer as Uint8Array
        n.fill(0x00)
        return n
    })
})

afterEach(() => {
    jest.spyOn(global.crypto, 'getRandomValues').mockRestore()
})

//https://github.com/trezor/python-mnemonic/blob/master/vectors.json


describe('Wallet', () => {

    it('Wallet encryption', async () => {
        process.env.I_AM_SURE_I_AM_NOT_IN_PRODUCTION = String(true)

        const wallet = await InitializeWallet.fromPhrase('', 'abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about')
        await wallet.encrypt('Fake password')
        expect(wallet.isEncrypted).toBeTruthy()

        await wallet.runUnencrypted('Fake password', async () => {
            const seed = await wallet.getSeed()
            expect(seed.hexa).toBe(
                '5eb00bbddcf069084889a8ab9155568165f5c453' +
                'ccb85e70811aaed6f6da5fc19a5ac40b389cd370' +
                'd086206dec8aa6c43daea6690f20ad3d8d48b2d2' +
                'ce9e38e4')
        })

        await expect(async () => wallet.Phrase).rejects.toThrow(WalletIsEncrypted)
    })



    it('Wallet from phrase (English)', async () => {
        process.env.I_AM_SURE_I_AM_NOT_IN_PRODUCTION = String(true)
        const wallet = await InitializeWallet.fromPhrase('',
            'abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about')

        const seed = await wallet.getSeed()
        expect(seed.hexa).toBe(
            '5eb00bbddcf069084889a8ab9155568165f5c453' +
            'ccb85e70811aaed6f6da5fc19a5ac40b389cd370' +
            'd086206dec8aa6c43daea6690f20ad3d8d48b2d2' +
            'ce9e38e4')
        expect(seed.xpriv).toBe(
            'xprv9s21ZrQH143K3GJpoapnV8SFfukcVBSfeCficPSGfubmSFDxo1kuHnLisriDvSnRRuL2Qrg5ggqHKNVpxR86QEC8w35uxmGoggxtQTPvfUu')
    })

    it('Wallet from phrase (Chinese)', async () => {
        process.env.I_AM_SURE_I_AM_NOT_IN_PRODUCTION = String(true)
        const wallet = await InitializeWallet.fromPhrase('',
            '的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 的 动')

        const seed = await wallet.getSeed()
        expect(seed.hexa).toBe(
            '87a89b5145f00cc0dee954940deec9c6606bc61a5d34f7c0e4fadb46ac35665317f47ca3f2e57f451dde0fd96c6a4d06a51c6fbb9317da2fa80ffa66dda3aefc')
        expect(seed.xpriv).toBe(
            'xprv9s21ZrQH143K33GSJBSLkyd6JaaKcwsVKMMRKrHJvh6q4UeSvoYs3nJaTkBywwSYHZ4SrZi5kTamxHWHJMaWZnrE3zP3vCXBa1ArMa8hJQq'
        )
    })

    it('Invalid phrase', async () => {
        await expect(async () => await InitializeWallet.fromPhrase('',
            'state small satisfied budge ant provision pause clue still shrink score chemistry')).rejects.toThrow(EncodingError)
    })


    it('Wallet from seed', async () => {
        process.env.I_AM_SURE_I_AM_NOT_IN_PRODUCTION = String(true)
        const wallet = await InitializeWallet.fromSeed('', '000102030405060708090a0b0c0d0e0f')
        expect(wallet.Seed.xpriv).toBe('xprv9s21ZrQH143K3QTDL4LXw2F7HEK3wJUD2nW2nRk4stbPy6cq3jPPqjiChkVvvNKmPGJxWUtg6LnF5kejMRNNU3TGtRBeJgk33yuGBxrMPHi')
    })

    it('Invalid seed', async () => {
        await expect(async () => await InitializeWallet.fromSeed('',
            'Invalid hexadecimal')).rejects.toThrow(EncodingError)
    })

    it('Wallet from private key', async () => {
        process.env.I_AM_SURE_I_AM_NOT_IN_PRODUCTION = String(true)
        const wallet = await InitializeWallet.fromPrivateKey('', '741745080050f2ce656aaa2a983a6b510caa706643e1ad05214feac6677ba657')
        expect((wallet.PrivateKey).hexa).toBe('741745080050f2ce656aaa2a983a6b510caa706643e1ad05214feac6677ba657')
    })

    it('Invalid private key', async () => {
        await expect(async () => await InitializeWallet.fromPrivateKey('',
            'Invalid hexadecimal')).rejects.toThrow(EncodingError)
    })

    it('New wallet', async () => {
        process.env.I_AM_SURE_I_AM_NOT_IN_PRODUCTION = String(true)
        const wallet = await InitializeWallet.create('')
        const phrase = wallet.Phrase
        expect(phrase).toBe('abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about')
    })

    it('New wallet with phrases in different languages', async () => {
        process.env.I_AM_SURE_I_AM_NOT_IN_PRODUCTION = String(true)

        let wallet = await InitializeWallet.create('', null)
        expect(wallet.Phrase).toBe(
            'abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about')

        wallet = await InitializeWallet.create('', 'english')
        expect(wallet.Phrase).toBe(
            'abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about')

        wallet = await InitializeWallet.create('', 'czech')
        expect(wallet.Phrase).toMatchSnapshot()

        wallet = await InitializeWallet.create('', 'french')
        expect(wallet.Phrase).toMatchSnapshot()

        wallet = await InitializeWallet.create('', 'italian')
        expect(wallet.Phrase).toMatchSnapshot()

        wallet = await InitializeWallet.create('', 'korean')
        expect(wallet.Phrase).toMatchSnapshot()

        wallet = await InitializeWallet.create('', 'japanese')
        expect(wallet.Phrase).toMatchSnapshot()

        wallet = await InitializeWallet.create('', 'portuguese')
        expect(wallet.Phrase).toMatchSnapshot()

        wallet = await InitializeWallet.create('', 'simplifiedChinese')
        expect(wallet.Phrase).toMatchSnapshot()

        wallet = await InitializeWallet.create('', 'spanish')
        expect(wallet.Phrase).toMatchSnapshot()

        wallet = await InitializeWallet.create('', 'traditionalChinese')
        expect(wallet.Phrase).toMatchSnapshot()
    })

    it('New wallet with different lengths', async () => {
        process.env.I_AM_SURE_I_AM_NOT_IN_PRODUCTION = String(true)

        let wallet = await InitializeWallet.create('', 'english', 12)
        expect(wallet.Phrase).toBe(
            'abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about')

        wallet = await InitializeWallet.create('', 'english', 15)
        expect(wallet.Phrase).toMatchSnapshot()

        wallet = await InitializeWallet.create('', 'english', 18)
        expect(wallet.Phrase).toMatchSnapshot()

        wallet = await InitializeWallet.create('', 'english', 21)
        expect(wallet.Phrase).toMatchSnapshot()

        wallet = await InitializeWallet.create('', 'english', 24)
        expect(wallet.Phrase).toMatchSnapshot()

    })

    it('Import wallet from keystore (phrase v1 format)', async () => {
        process.env.I_AM_SURE_I_AM_NOT_IN_PRODUCTION = String(true)

        const keystore = '{"version":1,"crypto":{"cipher":"aes-128-ctr","cipherparams":{"iv":"00bee8a13b05a67c1c5dcdeaf461ee4a"},"ciphertext":"e96757767288314beb670abc4eddc40f1a36fc6f5812594d15c022aad88aff4d3a998501884b752bca0967f4c2dc6d9dabf922fd056cefa8cbc5b04a7beb64ae4f93b4af4c7543c8e83699","kdf":"pbkdf2","kdfparams":{"c":262144,"prf":"hmac-sha256","dklen":32,"salt":"da46c078f9657ecec39e9bddee7def09f37e06025c1eceb2b9412454800814db"},"mac":"3e56e57d5fcc6873f9c09e29e1a76f6f1f1c8a534f51f164ef2064607fb136c6"}}'
        await expect(async () => await InitializeWallet.fromKeystore('', keystore, 'Incorrect password')).rejects.toThrow(IncorrectPassword)
        const wallet = await InitializeWallet.fromKeystore('', keystore, '1234')
        expect(await wallet.currencies.bitcoin.getAddress()).toBe('bc1qu3c79ddecg67uazedwng6s2u9zn79mhhumgngk')
    })

    it('Import wallet from keystore (geth format)', async () => {
        process.env.I_AM_SURE_I_AM_NOT_IN_PRODUCTION = String(true)

        const keystore = '{"address":"014c1c3a420c781061a8decc2f3e00cda1c59816","crypto":{"cipher":"aes-128-ctr","ciphertext":"6b35eafed8217cb2bda7d57f7e165e26caedb653defbf83deed4acbb3b47609e","cipherparams":{"iv":"be2065014801e9264e114c32728c7c09"},"kdf":"scrypt","kdfparams":{"dklen":32,"n":262144,"p":1,"r":8,"salt":"a109b4a90baa55fe59c47a9915de0dd91a357e1ba02b26234a5fe2b8cbcceca2"},"mac":"8905d186d71c6fd6e2df02f4bc3233a46f38aedba5dc55b6c4efe0fbf87979e6"},"id":"39a6290e-f719-4339-9b0f-97fb7ed3e4b4","version":3}'
        await expect(async () => await InitializeWallet.fromKeystore('', keystore, 'Incorrect password')).rejects.toThrow(IncorrectPassword)
        const wallet = await InitializeWallet.fromKeystore('', keystore, '1234')
        expect(await wallet.currencies.ethereum.getAddress()).toBe('0x014c1C3A420c781061A8DECc2f3E00cda1C59816')
    })

    it('Get all balances', async () => {
        process.env.I_AM_SURE_I_AM_NOT_IN_PRODUCTION = String(true)

        const wallet = await InitializeWallet.fromPhrase('',
            'abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon about')
        expect(await wallet.getAllBalances()).toMatchSnapshot()
    })

})
