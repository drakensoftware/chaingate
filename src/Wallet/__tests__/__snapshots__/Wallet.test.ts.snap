// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wallet Get all balances 1`] = `
[
  {
    "balance": {
      "confirmed": CurrencyAmount {
        "baseAmount": "0",
        "baseSymbol": "ARB",
        "minimalUnitAmount": "0",
        "minimalUnitSymbol": "wei",
      },
      "unconfirmed": CurrencyAmount {
        "baseAmount": "0",
        "baseSymbol": "ARB",
        "minimalUnitAmount": "0",
        "minimalUnitSymbol": "wei",
      },
    },
    "currency": {
      "decimals": 18,
      "defaultDerivationPath": "m/44'/60'/0'/0/0",
      "id": "arbitrum",
      "minimalUnitSymbol": "wei",
      "name": "Arbitrum One",
      "svgLogoUrl": "https://api.chaingate.dev/arbitrum/logo",
      "symbol": "ARB",
    },
  },
  {
    "balance": {
      "confirmed": CurrencyAmount {
        "baseAmount": "0",
        "baseSymbol": "AVAX",
        "minimalUnitAmount": "0",
        "minimalUnitSymbol": "wei",
      },
      "unconfirmed": CurrencyAmount {
        "baseAmount": "0",
        "baseSymbol": "AVAX",
        "minimalUnitAmount": "0",
        "minimalUnitSymbol": "wei",
      },
    },
    "currency": {
      "decimals": 18,
      "defaultDerivationPath": "m/44'/60'/0'/0/0",
      "id": "avalanche",
      "minimalUnitSymbol": "wei",
      "name": "Avalanche C-Chain",
      "svgLogoUrl": "https://api.chaingate.dev/avalanche/logo",
      "symbol": "AVAX",
    },
  },
  {
    "balance": {
      "confirmed": CurrencyAmount {
        "baseAmount": "2.083e-14",
        "baseSymbol": "BSC",
        "minimalUnitAmount": "20830",
        "minimalUnitSymbol": "wei",
      },
      "unconfirmed": CurrencyAmount {
        "baseAmount": "0",
        "baseSymbol": "BSC",
        "minimalUnitAmount": "0",
        "minimalUnitSymbol": "wei",
      },
    },
    "currency": {
      "decimals": 18,
      "defaultDerivationPath": "m/44'/60'/0'/0/0",
      "id": "binanceSmartChain",
      "minimalUnitSymbol": "wei",
      "name": "Binance Smart Chain",
      "svgLogoUrl": "https://api.chaingate.dev/binance/logo",
      "symbol": "BSC",
    },
  },
  {
    "balance": {
      "confirmed": CurrencyAmount {
        "baseAmount": "0",
        "baseSymbol": "BTC",
        "minimalUnitAmount": "0",
        "minimalUnitSymbol": "satoshi",
      },
      "unconfirmed": CurrencyAmount {
        "baseAmount": "0",
        "baseSymbol": "BTC",
        "minimalUnitAmount": "0",
        "minimalUnitSymbol": "satoshi",
      },
    },
    "currency": {
      "decimals": 8,
      "defaultDerivationPath": "m/84'/0'/0'/0/0",
      "id": "bitcoin",
      "minimalUnitSymbol": "satoshi",
      "name": "Bitcoin",
      "svgLogoUrl": "https://api.chaingate.dev/bitcoin/logo",
      "symbol": "BTC",
    },
  },
  {
    "balance": {
      "confirmed": CurrencyAmount {
        "baseAmount": "0",
        "baseSymbol": "BTC-TESTNET",
        "minimalUnitAmount": "0",
        "minimalUnitSymbol": "satoshi",
      },
      "unconfirmed": CurrencyAmount {
        "baseAmount": "0",
        "baseSymbol": "BTC-TESTNET",
        "minimalUnitAmount": "0",
        "minimalUnitSymbol": "satoshi",
      },
    },
    "currency": {
      "decimals": 8,
      "defaultDerivationPath": "m/84'/1'/0'/0/0",
      "id": "bitcoin-testnet",
      "minimalUnitSymbol": "satoshi",
      "name": "Bitcoin",
      "svgLogoUrl": "https://api.chaingate.dev/bitcoin/logo",
      "symbol": "BTC-TESTNET",
    },
  },
  {
    "balance": {
      "confirmed": CurrencyAmount {
        "baseAmount": "0",
        "baseSymbol": "BOBA",
        "minimalUnitAmount": "0",
        "minimalUnitSymbol": "wei",
      },
      "unconfirmed": CurrencyAmount {
        "baseAmount": "0",
        "baseSymbol": "BOBA",
        "minimalUnitAmount": "0",
        "minimalUnitSymbol": "wei",
      },
    },
    "currency": {
      "decimals": 18,
      "defaultDerivationPath": "m/44'/60'/0'/0/0",
      "id": "boba",
      "minimalUnitSymbol": "wei",
      "name": "Boba Network",
      "svgLogoUrl": "https://api.chaingate.dev/boba/logo",
      "symbol": "BOBA",
    },
  },
  {
    "balance": {
      "confirmed": CurrencyAmount {
        "baseAmount": "1.1909e-14",
        "baseSymbol": "ETH",
        "minimalUnitAmount": "11909",
        "minimalUnitSymbol": "wei",
      },
      "unconfirmed": CurrencyAmount {
        "baseAmount": "0",
        "baseSymbol": "ETH",
        "minimalUnitAmount": "0",
        "minimalUnitSymbol": "wei",
      },
    },
    "currency": {
      "decimals": 18,
      "defaultDerivationPath": "m/44'/60'/0'/0/0",
      "id": "ethereum",
      "minimalUnitSymbol": "wei",
      "name": "Ethereum",
      "svgLogoUrl": "https://api.chaingate.dev/ethereum/logo",
      "symbol": "ETH",
    },
  },
  {
    "balance": {
      "confirmed": CurrencyAmount {
        "baseAmount": "1.9e-14",
        "baseSymbol": "MATIC",
        "minimalUnitAmount": "19000",
        "minimalUnitSymbol": "wei",
      },
      "unconfirmed": CurrencyAmount {
        "baseAmount": "0",
        "baseSymbol": "MATIC",
        "minimalUnitAmount": "0",
        "minimalUnitSymbol": "wei",
      },
    },
    "currency": {
      "decimals": 18,
      "defaultDerivationPath": "m/44'/60'/0'/0/0",
      "id": "polygon",
      "minimalUnitSymbol": "wei",
      "name": "Polygon",
      "svgLogoUrl": "https://api.chaingate.dev/polygon/logo",
      "symbol": "MATIC",
    },
  },
]
`;

exports[`Wallet New wallet with different lengths 1`] = `"abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon address"`;

exports[`Wallet New wallet with different lengths 2`] = `"abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon agent"`;

exports[`Wallet New wallet with different lengths 3`] = `"abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon admit"`;

exports[`Wallet New wallet with different lengths 4`] = `"abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon abandon art"`;

exports[`Wallet New wallet with phrases in different languages 1`] = `"abdikace abdikace abdikace abdikace abdikace abdikace abdikace abdikace abdikace abdikace abdikace agrese"`;

exports[`Wallet New wallet with phrases in different languages 2`] = `"abaisser abaisser abaisser abaisser abaisser abaisser abaisser abaisser abaisser abaisser abaisser abeille"`;

exports[`Wallet New wallet with phrases in different languages 3`] = `"abaco abaco abaco abaco abaco abaco abaco abaco abaco abaco abaco abete"`;

exports[`Wallet New wallet with phrases in different languages 4`] = `"가격 가격 가격 가격 가격 가격 가격 가격 가격 가격 가격 가능"`;

exports[`Wallet New wallet with phrases in different languages 5`] = `"あいこくしん　あいこくしん　あいこくしん　あいこくしん　あいこくしん　あいこくしん　あいこくしん　あいこくしん　あいこくしん　あいこくしん　あいこくしん　あおぞら"`;

exports[`Wallet New wallet with phrases in different languages 6`] = `"abacate abacate abacate abacate abacate abacate abacate abacate abacate abacate abacate abater"`;

exports[`Wallet New wallet with phrases in different languages 7`] = `"的 的 的 的 的 的 的 的 的 的 的 在"`;

exports[`Wallet New wallet with phrases in different languages 8`] = `"ábaco ábaco ábaco ábaco ábaco ábaco ábaco ábaco ábaco ábaco ábaco abierto"`;

exports[`Wallet New wallet with phrases in different languages 9`] = `"的 的 的 的 的 的 的 的 的 的 的 在"`;
